sources/web/datalab/fileSearch.ts

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2017 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+/// <reference path="../../../third_party/externs/ts/node/node.d.ts" />
+/// <reference path="common.d.ts" />
+
+import http = require('http');
+import url = require('url');
+import fs = require('fs');
+import path = require('path');
+import logging = require('./logging');
+
+var appSettings: common.Settings;
+var fileIndex: string[] = [];
+var tooManyFiles: boolean = false;
+const fileCountLimit = 1000000;
+
+// this is matched by the client javascript to display a message that
+// there are more results than shown. Make sure these are in sync
+const clientResultSize = 20; 
+
+/**
+ * Implements the file search/filter request handling.
+ * @param request the incoming search request.
+ * @param response the outgoing search response.
+ */
+function requestHandler(request: http.ServerRequest, response: http.ServerResponse): void {
+  const parsedUrl = url.parse(request.url, true);
+  const pattern = parsedUrl.query['pattern'];  
+  const statusCheck = parsedUrl.query['status'];
+
+  response.writeHead(200, { 'Content-Type': 'application/json' });
+  if (pattern !== undefined) {
+    const results = filter(pattern);
+    response.write(JSON.stringify(results));
+  } else if (statusCheck !== undefined) {
+    response.write(JSON.stringify({
+      tooManyFiles: tooManyFiles
+    }));
+  }
+  response.end();
+}
+
+/**
+ * Builds an index of all files in the content directory to make search faster
+ */
+export function indexFiles(): void {
+  const startTime = process.hrtime();
+  index(appSettings.contentDir + '/', () => {
+    const indexTime = process.hrtime(startTime);
+    if (tooManyFiles) {
+      logging.getLogger().error('Hit file index maximum limit, searching files should be disabled');
+    } else {
+      logging.getLogger().info('Finished indexing ' + fileIndex.length + ' files in ' + indexTime[0] + ' seconds');
+    }
+  });
+}
+
+/**
+ * Recursively indexes all files under the given search path into the global fileIndex object
+ * @param searchpath the path prefix to start indexing
+ */
+function index(searchpath: string, callback: Function): void {
+  // stop the search if we've reached the file count limit
+  if (fileIndex.length >= fileCountLimit) {
```

Comment:
I don't think your limit logic is correct. This only checks at the beginning of a directory, but while scanning a directory it will continue to push files onto the list without checking its size.
Maybe test this code by setting fileCountLimit to a small number like 10 and see what happens.